modules/packages/apk

```diff
@@ -0,0 +1,91 @@
+#!/bin/sh -e
+get_package_data() {
+  name="${INPUT_File?File must be given to get-package-data}"
+  echo PackageType=repo
+  echo Name=$name
+}
+
+list_installed() {
+  # Example pkg output:
+  # Name         Version  Rev   Developer  Notes
+  # core         16-2.30  3748  canonical  core
+  # hello-world  6.3      27    canonical  -
+  #
+  # After rewrite:
+  # Name=core
+  # Version=16-2.30
+  # Architecture=none
+  apk list --installed | sed 's/-/ /' | awk '
+{
+    printf("Name=%s\n",$1)
+    printf("Version=%s\n",$2)
+    printf("Architecture=%s\n",$3)
+}'
+}
+
+repo_install() {
+  name="${INPUT_Name?Name must be given to repo-install}"
+  version="${INPUT_Version}"
+  if [ ! -z "${INPUT_Version}" ]; then
+    apk add "$name=$version" 2>/dev/null >/dev/null
+  else
+    apk add --upgrade "$name" 2>/dev/null >/dev/null
+  fi
+}
+
+
+list_updates() {
+# for some odd reason --upgradable does not work where -u does
+  apk list -u | sed 's/-/ /' | awk '
+{
+    printf("Name=%s\n",$1)
+    printf("Version=%s\n",$2)
+    printf("Architecture=%s\n",$3)
+}'
+}
+
+remove() {
+  name="${INPUT_Name?Name must be given to remove}"
+  apk del "$name" 2>/dev/null >/dev/null
+}
+
+main() {
+  command=$1
+
+  # Output maybe contain backslashes, and we don't want those to end up escaping
+  # something so we use use -r with read.
+  while read -r line; do
+    # Note that line is a variable assignment, e.g.
+    # INPUT_File=syncthing
+    export INPUT_$line
```

Comment:
so true. well done. We should mark this as an issue with `snap` module as well then, as this is a copy/paste/modify :(

Reply:
Interesting question though, is when does this actually happen?

Each packages promise can only specify one package name and each is processed individually as far as I can tell.

I'll take a look at the core C code a bit to see if I can tell when multiples would be sent.

Reply:
good point! Maybe docs need to be updated

Reply:
Good find @Lex-2008, @kacf might know about why it needs to handle mutlple.

Reply:
I *think* that it only plays a role if you are installing one package for multiple architectures (quite common in the i386->amd64 transition, but getting less and less common). The versions of these packages cannot ever conflict, so it will try to install both types at once. See [this section](https://github.com/cfengine/core/blob/master/cf-agent/package_module.c#L1232-L1242).

I think the theory was also that we could optimize by installing slists of packages, but I don't think this was ever implemented.

Reply:
So @Lex-2008 and @nickanderson do you want to move forward with as-is or implement supporting multiple packages?

Reply:
Right, since it wasn't implemented, but was working, before - it should be fine to keep it like this and maybe file a ticket for later.